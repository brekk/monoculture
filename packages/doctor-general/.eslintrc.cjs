module.exports = {
  root: true,
  extends: ['monoculture'],
  overrides: [
    {
      files: ['autotests/*'],
      rules: {
        'unused-imports/no-unused-vars': 0,
        '@typescript-eslint/no-unused-vars': 0,
        'import/no-duplicates': 0,
      },
    },
  ],
}
