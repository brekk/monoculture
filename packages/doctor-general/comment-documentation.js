import { dirname, join as pathJoin } from 'node:path'
import { isNotEmpty } from 'inherent'
import {
  defaultTo,
  filter,
  map,
  pathOr,
  pipe,
  prop,
  propOr,
  reduce,
} from 'ramda'
import { log } from './log'
import { stripRelative } from './text'
import { combineFiles } from './file'

export const parsePackageName = y => {
  const slash = y.indexOf('/')
  const start = slash + 1
  const end = y.indexOf('/', start)
  return y.slice(start, end)
}

export const filterAndStructureComments = pipe(
  log.doc('filter and structure!'),
  filter(pipe(propOr([], 'comments'), isNotEmpty)),
  map(raw => {
    log.doc('file', raw)
    const filename = stripRelative(raw.filename)
    return {
      ...raw,
      comments: raw.comments.map(r => ({ ...r, filename })),
      filename,
      workspace: parsePackageName(filename),
    }
  }),
  reduce((agg, file) => {
    const filenames = map(prop('filename'), agg)
    const alreadyInList = filenames.includes(file.filename)
    const anyFile = file.comments.filter(({ structure }) => structure.asFile)
    const someFile = anyFile.length > 0 ? anyFile[0] : false
    const asFilePath = pipe(
      defaultTo({}),
      pathOr('???', ['structure', 'asFile'])
    )(someFile)
    const withOrder = pipe(pathOr('0', ['structure', 'order']), x =>
      parseInt(x, 10)
    )(someFile)
    const dir = dirname(file.filename)
    const newFile = someFile ? pathJoin(dir, asFilePath) : '???'
    return alreadyInList
      ? map(raw => {
          const check = raw.filename === file.filename
          return check ? combineFiles(raw.order < withOrder, raw, file) : raw
        })(agg)
      : [
          ...agg,
          someFile
            ? {
                ...file,
                filename: newFile,
                order: withOrder,
                originalFilename: file.filename,
              }
            : file,
        ]
  }, [])
)
