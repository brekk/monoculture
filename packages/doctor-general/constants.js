export const MAGIC_IMPORT_KEY = 'drgen-import-above'
export const TESTABLE_EXAMPLE = 'test=true'
