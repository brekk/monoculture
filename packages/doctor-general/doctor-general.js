export * from './comment'
export * from './file'
export * from './parse'
export * from './renderer-markdown'
export * from './renderer-jest'
export * from './stats'
export * from './string'
export * from './text'
