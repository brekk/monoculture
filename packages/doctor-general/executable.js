import { drgen } from './cli'
import { fork } from 'fluture'

// fork it so it actually executes!
// eslint-disable-next-line no-console
fork(console.error)(console.log)(drgen(() => {}, process.argv))
