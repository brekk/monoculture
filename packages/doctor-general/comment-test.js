import { extname, basename, dirname, join as pathJoin } from 'node:path'
import { autobox, isNotEmpty } from 'inherent'
import { TESTABLE_EXAMPLE } from './constants'
import {
  ap,
  curry,
  defaultTo,
  filter,
  head,
  includes,
  map,
  pathOr,
  pipe,
  prop,
  propOr,
  reduce,
} from 'ramda'
import { stripRelative } from './text'
import { combineFiles } from './file'

export const hasTestableExample = pipe(
  pathOr('', ['structure', 'example']),
  includes(TESTABLE_EXAMPLE)
)

export const filterAndStructureTests = pipe(
  filter(pipe(propOr([], 'comments'), filter(hasTestableExample), isNotEmpty)),
  map(raw => {
    const filename = stripRelative(raw.filename)
    const ext = extname(filename)
    return {
      ...raw,
      comments: raw.comments.map(r => ({ ...r, filename })),
      filename,
      testPath: `${basename(filename, ext)}.spec${ext}`,
    }
  }),
  reduce((agg, file) => {
    const filenames = map(prop('filename'), agg)
    const alreadyInList = filenames.includes(file.filename)
    const anyFile = file.comments.filter(({ structure }) => structure.asFile)
    const someFile = anyFile.length > 0 ? anyFile[0] : false
    const asFilePath = pipe(
      defaultTo({}),
      pathOr('???', ['structure', 'asFile'])
    )(someFile)
    const withOrder = pipe(pathOr('0', ['structure', 'order']), x =>
      parseInt(x, 10)
    )(someFile)
    const dir = dirname(file.filename)
    const newFile = someFile ? pathJoin(dir, asFilePath) : '???'
    return alreadyInList
      ? map(raw => {
          const check = raw.filename === file.filename
          return check ? combineFiles(raw.order < withOrder, raw, file) : raw
        })(agg)
      : [
          ...agg,
          someFile
            ? {
                ...file,
                filename: newFile,
                order: withOrder,
                originalFilename: file.filename,
              }
            : file,
        ]
  }, [])
)
export const getImportsForTests = curry((testMode, file) =>
  !testMode
    ? []
    : pipe(
        map(
          pipe(
            autobox,
            ap([pathOr(false, ['structure', 'name']), hasTestableExample])
          )
        ),
        filter(([a, b]) => a && b),
        map(head)
      )(file.comments)
)
