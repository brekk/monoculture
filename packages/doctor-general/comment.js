import { join as pathJoin } from 'node:path'
import { parallel } from 'fluture'
import { isNotEmpty, autobox } from 'inherent'
import { TESTABLE_EXAMPLE } from './constants'
import {
  prop,
  __ as $,
  addIndex,
  always as K,
  any,
  ap,
  applySpec,
  both,
  chain,
  concat,
  cond,
  curry,
  defaultTo,
  either,
  equals,
  filter,
  findIndex,
  flatten,
  groupBy,
  head,
  identity as I,
  ifElse,
  includes,
  is,
  last,
  length,
  map,
  match,
  pathOr,
  pipe,
  propOr,
  reduce,
  reject,
  replace,
  slice,
  split,
  startsWith,
  test as testRegExp,
  toPairs,
  trim,
  uniq,
  when,
} from 'ramda'
import { writeFileWithAutoPath } from 'file-system'
import { unlines } from 'knot'

import { log } from './log'
import { findJSDocKeywords, cleanupKeywords } from './file'
import { formatComment, trimComment, wipeComment } from './text'

/**
 * Check to see if a comment has an example within its structure.
 * @name hasExample
 * @example
 * ```js test=true
 * expect(hasExample({structure: {example: [`test=true`]}})).toBeTruthy()
 * expect(hasExample({})).toBeFalsy()
 * expect(hasExample('zipzop')).toBeFalsy()
 * ```
 */
export const hasExample = ifElse(
  is(Object),
  pipe(pathOr('', ['structure', 'example']), includes(TESTABLE_EXAMPLE)),
  K(false)
)

/**
 * Pull all imports from a given file, including `@curried` examples
 * @name getImportsForTests
 * @example
 * ```js test=true
 * const file = { comments: [
 *   {
 *     structure: { curried: [
 *       { name: 'coolWithConfig', lines: ['test=true'] },
 *       { name: 'cool', lines: ['test=true'] }
 *     ] }
 *   },
 *   {
 *     structure: { name: 'otherFunc', example: ['test=true'] }
 *   }
 * ] }
 *
 * expect(getImportsForTests(file)).toEqual(['otherFunc', 'coolWithConfig', 'cool'])
 * ```
 */
export const getImportsForTests = pipe(
  propOr([], 'comments'),
  map(
    pipe(
      autobox,
      ap([
        pathOr(false, ['structure', 'name']),
        hasExample,
        pathOr([], ['structure', 'curried']),
      ])
    )
  ),
  reduce(function handleCurriedExamples(agg, [a, b, c]) {
    return pipe(
      concat(
        isNotEmpty(c)
          ? map(function _testCurriedExample({ name, lines: example }) {
              return [name, includes(TESTABLE_EXAMPLE)(example)]
            })(c)
          : [[a, b]]
      )
    )(agg)
  }, []),
  filter(([a, b]) => a && b),
  map(head)
)

export const LINK_REGEX = /\{@link (.*)+\}/g

/**
 * @name matchLinks
 * @example
 * ```js test=true
 * expect(matchLinks([])).toEqual([])
 * expect(matchLinks([`{@link cool}`])).toEqual(['cool'])
 * ```
 */
export const matchLinks = pipe(
  chain(match(LINK_REGEX)),
  map(z => slice('{@link '.length, z.length - 1, z))
)
const ORDERED_LIST_ITEM = /^\s\d*\.\s/g
const CURRIED_LIST_ITEM = /^\d+\.\s(\w+)\s-\s(.*)/g
const isListItem = testRegExp(ORDERED_LIST_ITEM)

export const getCurriedDefinition = curry(
  function _getCurriedDefinition(file, end, i) {
    return pipe(
      slice(i + 1, end),
      map(
        pipe(trimComment, when(pipe(trim, equals('*')), K('')), line => [
          isListItem(line),
          line,
        ])
      ),
      reduce(
        function findingCurriedDefinitions(
          { subject, lines, defs },
          [isDefinition, content]
        ) {
          return isDefinition
            ? {
                defs: lines.length ? [...defs, { lines, subject }] : defs,
                lines: [],
                subject: content,
              }
            : {
                subject,
                defs,
                lines: [
                  ...lines,
                  content.startsWith('    ') ? content.slice(4) : content,
                ],
              }
        },
        { subject: null, lines: [], defs: [] }
      ),
      function handleFinalDefinition({ subject, defs, lines }) {
        return [...defs, { lines, subject }]
      },
      map(function reformatCurriedDefs({ lines, subject }) {
        const matched = subject
          ? trim(subject).replace(CURRIED_LIST_ITEM, '$1⩇$2')
          : ''
        const [name, summary] = split('⩇', matched)
        return {
          name,
          summary,
          lines: pipe(reject(equals('@example')), unlines)(lines),
        }
      })
    )(file)
  }
)

/**
 * Grab the summary from raw lines, given some indices to slice
 * @name getPageSummary
 * @example
 * ```js test=true
 * const rawLines = [
 *   ' * @pageSummary Hey cool this is a multi-line',
 *   ' * description of stuff in the whole file',
 *   ' * @page testPageSummary',
 *   ' * @huh notSure'
 * ]
 * expect(
 *   getPageSummary(rawLines, Infinity, 0)
 * ).toEqual([
 *   'Hey cool this is a multi-line',
 *   'description of stuff in the whole file'
 * ])
 * expect(
 *   getPageSummary([' * hey', ' * there'], Infinity, 0)
 * ).toEqual(['hey', 'there'])
 * ```
 */
export const getPageSummary = (rawLines, end, i) =>
  pipe(
    slice(i, end),
    map(pipe(trimComment, replace(/^\*$/g, ''))),
    filter(trim),
    map(replace(/@pageSummary\s/g, '')),
    function pageExclusion(lines) {
      // handle @page being thrown into @pageSummary
      const ex = findIndex(any(startsWith('@')), lines)
      return ex > -1 ? slice(0, ex, lines) : lines
    }
  )(rawLines)

/**
 * Is it an asterisk with maybe some whitespace around it?
 * @name isAsterisky
 * @example
 * ```js test=true
 * expect(isAsterisky('     * ')).toBeTruthy()
 * expect(isAsterisky('*')).toBeTruthy()
 * expect(isAsterisky('obelisk')).toBeFalsy()
 * ```
 */
export const isAsterisky = both(is(String), pipe(trim, equals('*')))

/**
 * @name stripEmptyCommentLines
 * @example
 * ```js test=true
 * expect(stripEmptyCommentLines('     *')).toEqual('')
 * expect(stripEmptyCommentLines('hooray!')).toEqual('hooray!')
 */
export const stripEmptyCommentLines = when(isAsterisky, K(''))

/*
 * Grab the example from raw lines and some indices to slice
 * @name getExample
 * @signature List String -> Integer -> Integer -> String
 * @example
 * ```js test=true
 * const rawLines = [
 *   '* Ouroboric summary',
 *   '* @name getExample',
 *   '* @example',
 *   '* ```js',
 *   '* getExample(rawLines)',
 *   '* ```',
 * ]
 * expect(getExample(rawLines, 5, 3)).toEqual('getExample(rawLines)')
 * ```
 */
export const getExample = curry(function _getExample(rawLines, end, i) {
  return pipe(
    slice(i + 1, end),
    map(pipe(trimComment, stripEmptyCommentLines)),
    unlines
  )(rawLines)
})

export const handleSpecificKeywords = curry(
  function _handleSpecificKeywords(keyword, value, rest, rawLines, end, i) {
    log.comment('RAW SPEC', { keyword, value, rest, rawLines, end, i })
    // there should only be three(?) kinds of tags
    // 1. single-line - ` * @blah zipzap zipzop` / ` * @blah {aFormat} [toParse] no newline yet`
    // 2. multi-line - ` * @heyYou it is actually
    //  * not syllable based, you know
    //  * cutting lines are of import` -- these are left associative and
    //    stop at the beginning of the first tag which starts a line
    // 3. ephemeral - these tags do specific "magical" things,
    //    like `@group` / `@page` / `@pageSummary` / `@addTo` / `@curried`
    //    and some of these tags then are not present in the resulting structure
    return pipe(
      cond([
        [equals('exported'), () => true],
        [equals('pageSummary'), () => getPageSummary(rawLines, end, i)],
        // curried function definitions afford named variants of the same function
        // see rawLines-system/fs.js for an example
        [equals('curried'), () => getCurriedDefinition(rawLines, end, i)],
        // if example found, pull from raw rawLines
        [equals('example'), () => getExample(rawLines, end, i)],
        // if see found, do some light cleanup
        [equals('see'), () => pipe(head, slice(0, -1))(rest)],
        // Consume pages + names as sentences
        [
          either(equals('name'), equals('page')),
          () => trim(`${value} ${rest.join(' ')}`),
        ],
        // if an array value, concat it
        [() => rest.length, () => [value, ...rest]],
        // otherwise just return the value
        [K(true), () => value],
      ])
    )(keyword)
  }
)

/**
 * @name structureComment
 * @example
 * ```js test=true
 * expect(structureComment([0, ' * @cool nice yes'])).toEqual('???')
 * ```
 */
export function structureComment([i, line]) {
  return [
    i,
    ifElse(
      pipe(findJSDocKeywords, isNotEmpty),
      pipe(wipeComment, cleanupKeywords),
      K(false)
    )(line),
  ]
}

const stripLeadingComment = pipe(
  trim,
  when(equals('/**'), K('')),
  when(equals('*/'), K('')),
  when(startsWith('*'), pipe(slice(1, Infinity), trim))
)

function uncommentBlock(block) {
  return map(([lineNum, line]) => [lineNum, stripLeadingComment(line)], block)
}

const segmentBlock = pipe(
  reduce(
    function segmentLogicalGroupsOfCommentBlock(agg, [lineNum, line]) {
      const lineParts = line.split(' ')
      const { structure } = agg
      const { current } = agg
      // eslint-disable-next-line prefer-const
      let [tag, ...additional] = lineParts
      const hasTag = startsWith('@', tag)
      if (!hasTag) {
        additional = lineParts
      }
      const cleanTag = hasTag ? tag.slice(1) : tag
      log.comment(
        `!!! ${cleanTag} -> (current: ${current}, hasTag: ${hasTag})`,
        additional
      )
      log.comment(`-> current`, structure[current])
      const currentStructure = current ? structure?.[current] : []
      const toAdd = additional.join(' ')
      if (hasTag) {
        const insert = additional.length ? toAdd : true
        const prev = structure?.[cleanTag] ?? false
        return {
          structure: {
            ...structure,
            [cleanTag]: prev
              ? // arrayify if there's a prior entry
                [...autobox(prev), insert]
              : insert,
          },
          current: cleanTag,
        }
      }

      if (!current) {
        return {
          structure: {
            ...structure,
            description: [toAdd],
          },
          current: 'description',
        }
      }
      return {
        structure: {
          ...structure,
          [current]: [
            ...(Array.isArray(currentStructure) ? currentStructure : []),
            toAdd,
          ],
        },
        current,
      }
    },
    { structure: {}, current: false }
  ),
  prop('structure'),
  z => ({ ...z, description: z.description.join(' ') })
)

// structureKeywords :: List String -> CommentBlock -> Integer -> CommentStructure
export const structureKeywords = curry(
  function _structureKeywords(file, block, end) {
    return pipe(
      uncommentBlock,
      filter(pipe(last, isNotEmpty)),
      segmentBlock
      /*
      map(function _handleSpecificRemap([i, [keyword, value = true, ...rest]]) {
        return [
          i,
          [keyword, handleSpecificKeywords(keyword, value, rest, file, end, i)],
        ]
      }),
      */
    )(block)
  }
)

const summarize = lines => {
  const stripped = reject(equals('*'), lines)
  return pipe(
    addIndex(map)((x, i) => ifElse(startsWith('@'), K(i), K(false))(x)),
    filter(I),
    head,
    defaultTo(length(stripped)),
    slice(0, $, stripped),
    unlines
  )(stripped)
}

// getFileGroup :: String -> Comment
const getFileGroup = propOr('', 'group')
const addTo = propOr('', 'addTo')

// objectifyComments :: Boolean -> String -> List Comment -> List CommentBlock
export const objectifyComments = curry(
  function _objectifyComments(filename, file, comments) {
    return reduce(
      (agg, block) =>
        agg.concat(
          pipe(
            // pass one
            applySpec({
              start: pipe(head, head),
              end: pipe(last, head),
              lines: formatComment,
            }),
            // pass two
            gen => {
              const structure = structureKeywords(file, block, gen.end)
              if (structure.page && !structure.name) {
                structure.name = structure.page
                structure.detail = gen.start
              }
              return {
                ...gen,
                summary: summarize(gen.lines),
                links: matchLinks(gen.lines),
                fileGroup: getFileGroup(structure),
                addTo: addTo(structure),
                structure,
                keywords: pipe(unlines, findJSDocKeywords, uniq, z => z.sort())(
                  gen.lines
                ),
              }
            }
          )(block)
        ),
      [],
      comments
    )
  }
)

const renderFileWith = curry(function _renderFileWith(
  { renderer, postRender },
  file
) {
  const info = { imports: getImportsForTests(file), file }
  return pipe(
    // XXX: we can't be tacit with this or we end up
    // forcing our processor to remember to be curried
    // which is a PITA to debug downstream
    map(function applyProcessorRenderer(raw) {
      return renderer(info, raw)
    }),
    log.verbose('rendered'),
    function applyProcessorPostRender(raw) {
      return postRender(info, raw)
    },
    log.verbose('postRendered')
  )(file.comments)
})

// hard to test until we set up a whole fixture + clean script
export const writeCommentsToFiles = curry(function _writeCommentsToFiles(
  { processor, outputDir },
  x
) {
  const { output: $outputPath, postProcess: $postProcess = (_a, _b, c) => c } =
    processor
  return pipe(
    toPairs,
    map(([workspace, commentedFiles]) => {
      const filesToWrite = map(file => {
        const filePathToWrite = pathJoin(
          outputDir,
          workspace,
          // this part is the structure of the file we wanna write
          $outputPath(file)
        )
        return pipe(
          renderFileWith(processor),
          writeFileWithAutoPath(filePathToWrite)
        )(file)
      })(commentedFiles)
      return $postProcess(
        { outputDir, workspace },
        commentedFiles,
        filesToWrite
      )
    }),
    flatten,
    parallel(10)
  )(x)
})

export const renderComments = curry(
  function _renderComments(processor, outputDir, x) {
    return chain(
      pipe(
        groupBy(propOr('unknown', processor.group)),
        writeCommentsToFiles({ processor, outputDir })
      )
    )(x)
  }
)

export const processComments = curry(
  function _processComments(bad, processor, x) {
    try {
      return processor.process(x)
    } catch (e) {
      bad(e)
    }
  }
)
