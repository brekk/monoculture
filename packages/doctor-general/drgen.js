import { join as pathJoin, dirname } from 'node:path'
import { signal } from 'kiddo'
import { log } from './log'
import {
  curry,
  unless,
  when,
  always as K,
  chain,
  identity as I,
  ifElse,
  map,
  pipe,
} from 'ramda'
import { parallel, Future } from 'fluture'
import { relativePathJoin } from 'file-system'
import { parseFile } from './parse'
import { renderComments, processComments } from './comment'
import { monorepoRunner } from './reader'
import { writeArtifact } from './writer'
import { validate, interrogate } from './interpreter'
export {
  validate as validateInterpreter,
  interrogate as checkInterpreter,
} from './interpreter'

const handleInvalidInterpreter = pipe(
  interrogate,
  ({ incorrectFields }) =>
    `Interpreter is invalid. Incorrect fields: ${incorrectFields}`
)

export const readAndProcessFiles = curry(function _readAndProcessFiles(
  cancel,
  config,
  { outputDir, relativeArtifact, relative },
  raw
) {
  return new Future((bad, good) => {
    const { interpreter, debug, input, output, artifact = false } = config
    return pipe(
      map(pipe(map(relative), chain(parseFile(debug)))),
      chain(parallel(10)),
      signal(cancel, {
        text: 'Parsing files...',
        successText: 'Parsed files!',
        failText: 'Unable to parse files!',
      }),
      map(processComments(bad, interpreter)),
      when(K(artifact), writeArtifact(relativeArtifact)),
      renderComments(interpreter, outputDir),
      signal(cancel, {
        text: 'Rendering comments...',
        successText:
          artifact || output
            ? `Wrote to${
                output
                  ? ' output: "' + output + '"' + (artifact ? ';' : '')
                  : ''
              }${artifact ? ' artifact: "' + artifact + '"' : ''}.`
            : `Processed ${input.join(' ')}`,
        failText: 'Unable to render comments!',
      }),
      map(K('')),
      good
    )(raw)
    return cancel
  })
})

export const drgen = curry(function _drgen(cancel, config) {
  return new Future(function _drgenF(bad, good) {
    const {
      interpreter,
      input,
      output,
      search: searchGlob,
      ignore,
      artifact = false,
      monorepo: monorepoMode = false,
      showMatchesOnly = false,
      cwd,
    } = config
    log.core('showMatchesOnly', showMatchesOnly)
    log.core('interpreter!', interpreter)
    if (!validate(interpreter)) {
      pipe(handleInvalidInterpreter, bad)(interpreter)
      return cancel
    }
    log.core('input', input)
    log.core('monorepoMode', monorepoMode)
    const rel = relativePathJoin(cwd)
    const outputDir = rel(output)
    const relativeArtifact = artifact ? rel(artifact) : false
    const relativeInput = map(rel, input)

    // TODO: we ought to segment the monorepoMode out further
    const pkgJson = monorepoMode ? relativeInput[0] : 'NOT RELEVANT'
    log.core('relating...', `${cwd} -> ${output}`)
    const root = pkgJson.slice(0, pkgJson.lastIndexOf('/'))
    const toLocal = map(dirname, input)
    console.log('root', root, 'HUH?', toLocal, 'input', input)
    return pipe(
      log.core(`monorepoMode?`),
      ifElse(
        I,
        () =>
          monorepoRunner(
            cancel,
            { showMatchesOnly, searchGlob, ignore },
            root,
            pkgJson
          ),
        () => good(input)
      ),
      unless(
        () => showMatchesOnly,
        chain(
          readAndProcessFiles(cancel, config, {
            outputDir,
            relativeArtifact,
            relative: r => (monorepoMode ? pathJoin(toLocal[0], r) : r),
          })
        )
      )
    )(monorepoMode)
    return cancel
  })
})
