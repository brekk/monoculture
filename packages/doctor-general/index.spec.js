import * as DR_GEN from './doctor-general'

test('default export', () => {
  expect(DR_GEN).toMatchSnapshot()
})
