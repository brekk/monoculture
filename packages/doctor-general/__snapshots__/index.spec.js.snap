// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default export 1`] = `
{
  "LINK_REGEX": /\\\\\\{@link \\(\\.\\*\\)\\+\\\\\\}/g,
  "MAGIC_COMMENT_END": "*/",
  "MAGIC_COMMENT_START": "/**",
  "MAGIC_IMPORT_KEY": "drgen-import-above",
  "TESTABLE_EXAMPLE": "test=true",
  "WHITESPACE_PRESERVING_TAGS": [
    "example",
    "curried",
  ],
  "addLineNumbers": [Function],
  "capitalToKebab": [Function],
  "checkInterpreter": [Function],
  "cleanupKeywords": [Function],
  "drgen": [Function],
  "findJSDocKeywords": [Function],
  "formatComment": [Function],
  "getExample": [Function],
  "getImportsForTests": [Function],
  "getPageSummary": [Function],
  "groupContiguousBlocks": [Function],
  "hasExample": [Function],
  "isAsterisky": [Function],
  "isJSDocComment": [Function],
  "iterateOverWorkspacesAndReadFiles": [Function],
  "log": {
    "comment": [Function],
    "config": [Function],
    "core": [Function],
    "doc": [Function],
    "error": [Function],
    "file": [Function],
    "interpreter": [Function],
    "parse": [Function],
    "reader": [Function],
    "renderer": [Function],
    "stats": [Function],
    "string": [Function],
    "text": [Function],
    "validate": [Function],
    "verbose": [Function],
  },
  "matchLinks": [Function],
  "monorepoRunner": [Function],
  "nixKeyword": [Function],
  "objectifyAllComments": [Function],
  "objectifyComments": [Function],
  "parse": [Function],
  "parseFile": [Function],
  "processComments": [Function],
  "processCurriedComment": [Function],
  "readAndProcessFiles": [Function],
  "readPackageJsonWorkspaces": [Function],
  "renderComments": [Function],
  "renderFileWith": [Function],
  "safeTrim": [Function],
  "segmentBlock": [Function],
  "stripEmptyCommentLines": [Function],
  "stripLeadingComment": [Function],
  "stripLeadingHyphen": [Function],
  "stripRelative": [Function],
  "structureKeywords": [Function],
  "trimComment": [Function],
  "trimSummary": [Function],
  "uncommentBlock": [Function],
  "validateInterpreter": [Function],
  "wipeComment": [Function],
  "writeArtifact": [Function],
  "writeCommentsToFiles": [Function],
}
`;
